### PR TITLE
[FIX] website_sale_digital: let salesman download digital via portal

### PR DESCRIPTION
**Pull request by yelizariev**
Portal shows all Sale Orders available for current user. For example, salesman can see his sales. If such a user can download digital files via product form in backend, it makes sense to let user download them via SO page on portal. However, it wasn't the case because /my/download requires product be purchased by current user [1]. Fix it by checking read access first.

STEPS

* in backend create SO with digital product (customer must be different from current user)
* create invoice and register a payment
* navigate to portal  (without using customer's token),
* open SO, click download on digital product

[1]: https://github.com/odoo/odoo/blob/1a24477fab4dd323cf94c010321d8942fb2c1a01/addons/website_sale_digital/models/account_invoice.py#L14-L22

opw-3144600

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Review by Demesmaeker — COMMENTED** [review]
Hello!

We don't exactly like the if... pass. Wouldn't it be better to make the access check in a variable and use that in the condition instead? It'd be clearer IMO

File: addons/website_sale_digital/controllers/main.py [hunk 1]
```diff
@@ -86,7 +86,10 @@ def download_attachment(self, attachment_id):
         res_id = attachment['res_id']
         purchased_products = request.env['account.move.line'].get_digital_purchases()
 
-        if res_model == 'product.product':
+        if request.env['ir.attachment'].search([('id', '=', int(attachment_id))]):
```
Comment (AntoineVDV) [inline addons/website_sale_digital/controllers/main.py:89]: I, too, find it confusing to skip the redirects with a `if: pass`.

Also, it's important that the security check is obvious to the reader. Here, it's not immediate that `.sudo()` is omitted on purpose from the `search` call while it was necessary with the `search_read` call. Instead, could you please use `check_access_{rights,rules}` to make the check obvious?

**Review by AntoineVDV — CHANGES_REQUESTED** [review]

**Comment by yelizariev** [comment]
@Demesmaeker @AntoineVDV 

I made explicit check for reading access.

I got rid of `if: pass`, but it requires a bigger diff. Let me know if you have a better idea

File: addons/website_sale_digital/controllers/main.py [hunk 2]
```diff
@@ -84,20 +85,30 @@ def download_attachment(self, attachment_id):
...
 
-        if res_model == 'product.product':
-            if res_id not in purchased_products:
-                return redirect(self.orders_page)
+        has_access = False
+        try:
+            self.env['ir.attachment'].browse(attachment_id).check('read')
+            has_access = True
+        except AccessError:
+            pass
```
[the remark this answers is not in the document]
Comment (Demesmaeker) [reply addons/website_sale_digital/controllers/main.py:94]: Another solution would be to put this in the except directly and take off the `has_read_access` variable.
This'd make a bigger diff, but it'd also be coherent. Both are fine IMO.

**Comment by AntoineVDV** [comment]
@robodoo delegate=Demesmaeker

**Review by AntoineVDV — APPROVED** [review]

**Comment by Demesmaeker** [comment]
Hello @yelizariev :slightly_smiling_face: 
Since it was easier to force-push than explain what we had in mind, we directly did it. Have a nice day =]

**Pull request closed without merging**